rust/src/pop3/pop3.rs

```diff
@@ -0,0 +1,498 @@
+        state.parse_response(buf, flow)
+    }
+}
+
+unsafe extern "C" fn rs_pop3_state_get_tx(state: *mut c_void, tx_id: u64) -> *mut c_void {
+    let state = cast_pointer!(state, POP3State);
+    match state.get_tx(tx_id) {
+        Some(tx) => {
+            return tx as *const _ as *mut _;
+        }
+        None => {
+            return std::ptr::null_mut();
+        }
+    }
+}
+
+unsafe extern "C" fn rs_pop3_state_get_tx_count(state: *mut c_void) -> u64 {
+    let state = cast_pointer!(state, POP3State);
+    return state.tx_id;
+}
+
+unsafe extern "C" fn rs_pop3_tx_get_alstate_progress(tx: *mut c_void, direction: u8) -> c_int {
+    let tx = cast_pointer!(tx, POP3Transaction);
+    if direction == Direction::ToServer as u8 {
+        (tx.command.is_some() || tx.complete) as c_int
+    } else {
+        (tx.response.is_some() || tx.complete) as c_int
+    }
+}
+
+export_tx_data_get!(pop3_get_tx_data, POP3Transaction);
+export_state_data_get!(pop3_get_state_data, POP3State);
+
+// Parser name as a C style string.
+const PARSER_NAME: &[u8] = b"pop3\0";
+
+#[no_mangle]
+pub unsafe extern "C" fn SCRegisterPop3Parser() {
+    let default_port = CString::new("[110]").unwrap();
+    let parser = RustParser {
+        name: PARSER_NAME.as_ptr() as *const c_char,
+        default_port: default_port.as_ptr(),
+        ipproto: IPPROTO_TCP,
+        probe_ts: Some(rs_pop3_probing_parser),
+        probe_tc: Some(rs_pop3_probing_parser),
+        min_depth: 0,
+        max_depth: 16,
+        state_new: rs_pop3_state_new,
+        state_free: rs_pop3_state_free,
+        tx_free: rs_pop3_state_tx_free,
+        parse_ts: rs_pop3_parse_request,
+        parse_tc: rs_pop3_parse_response,
+        get_tx_count: rs_pop3_state_get_tx_count,
+        get_tx: rs_pop3_state_get_tx,
+        tx_comp_st_ts: 1,
+        tx_comp_st_tc: 1,
+        tx_get_progress: rs_pop3_tx_get_alstate_progress,
+        get_eventinfo: Some(POP3Event::get_event_info),
+        get_eventinfo_byid: Some(POP3Event::get_event_info_by_id),
+        localstorage_new: None,
+        localstorage_free: None,
+        get_tx_files: None,
+        get_tx_iterator: Some(state_get_tx_iterator::<POP3State, POP3Transaction>),
+        get_tx_data: pop3_get_tx_data,
+        get_state_data: pop3_get_state_data,
+        apply_tx_config: None,
+        flags: APP_LAYER_PARSER_OPT_ACCEPT_GAPS,
+        get_frame_id_by_name: None,
+        get_frame_name_by_id: None,
+    };
+
+    let ip_proto_str = CString::new("tcp").unwrap();
+
+    if AppLayerProtoDetectConfProtoDetectionEnabled(ip_proto_str.as_ptr(), parser.name) != 0 {
+        let alproto = AppLayerRegisterProtocolDetection(&parser, 1);
+        ALPROTO_POP3 = alproto;
+        if AppLayerParserConfParserEnabled(ip_proto_str.as_ptr(), parser.name) != 0 {
+            let _ = AppLayerRegisterParser(&parser, alproto);
+        }
+        let retval = conf_get("app-layer.protocols.pop3.stream-depth");
```

Comment:
Why did you add this config option ?

Reply:
I think it makes sense that you'd want to be able to see all of the stream, just like with smb.
In fact we should probably add this option to all parsers.

Reply:
Could be done in `AppLayerRegisterParser` for all rust parsers ;-)